### PR TITLE
try benchmarking on 281 image dataset (Olsson Palace of Fine Arts)

### PR DESCRIPTION
**Pull request by johnwlambert**
Parallelize the CI to run 7 benchmarks at once, instead of 3 successively. Speeds up CI by X percent.

Also, add a new benchmark for 281 images, although with only a 20-frame lookahead for now (to be extended later).

### CODE REVIEW

**Comment by travisdriver** [comment]
How many images are in this dataset?

**Comment by johnwlambert** [comment]
> How many images are in this dataset?

@travisdriver  there are 281 in this dataset. But the baselines are tiny and the dataset is pretty easy. But the amount of images is a big stress test.

from https://www1.maths.lth.se/matematiklth/personal/calle/dataset/dataset.html

**Comment by travisdriver** [comment]
> > How many images are in this dataset?
> 
> @travisdriver there are 281 in this dataset. But the baselines are tiny and the dataset is pretty easy. But the amount of images is a big stress test.
> 
> from https://www1.maths.lth.se/matematiklth/personal/calle/dataset/dataset.html

Sounds good; the 16 image dataset that was posted on Slack had the same name so wasn't sure. Think this is a great benchmark to have.

**Comment by johnwlambert** [comment]
cc'ing @adisingh50 on the Palace-of-Fine-Arts-281 dataset

**Comment by akshay-krishnan** [comment]
Can we remove the benchmark for Door if its not giving us useful signals anymore?

File: .github/workflows/benchmark.yml [hunk 1]
```diff
@@ -38,64 +46,83 @@ jobs:
...
+              --config_name ${CONFIG_NAME}.yaml \
+              --image_extension $IMAGE_EXTENSION
+
+          elif [ "$LOADER_NAME" == "colmap-loader" ]; then
+              python gtsfm/runner/run_scene_optimizer_colmaploader.py \
+              --images_dir ${IMAGES_DIR} \
+              --colmap_files_dirpath $COLMAP_FILES_DIRPATH \
+              --max_frame_lookahead $MAX_FRAME_LOOKAHEAD \
+              --config_name ${CONFIG_NAME}.yaml
+          fi
```
Comment (akshay-krishnan) [inline .github/workflows/benchmark.yml:121]: Could you put this into a shell script and just call it here?

Comment (akshay-krishnan) [reply .github/workflows/benchmark.yml:121]: All the commands for that action I mean

Comment (johnwlambert) [reply .github/workflows/benchmark.yml]: Good call @akshay-krishnan. It's updated now.

File: .github/scripts/execute_single_benchmark.sh [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  --config_name ${CONFIG_NAME}.yaml \
+  --image_extension $IMAGE_EXTENSION
+
+elif [ "$LOADER_NAME" == "colmap-loader" ]; then
+  python gtsfm/runner/run_scene_optimizer_colmaploader.py \
+  --images_dir ${IMAGES_DIR} \
+  --colmap_files_dirpath $COLMAP_FILES_DIRPATH \
+  --max_frame_lookahead $MAX_FRAME_LOOKAHEAD \
+  --config_name ${CONFIG_NAME}.yaml
+fi
```
Comment (akshay-krishnan) [inline .github/scripts/execute_single_benchmark.sh:75]: We should have default else for all these elifs which prints an error and returns without executing the rest

Comment (johnwlambert) [reply .github/scripts/execute_single_benchmark.sh:75]: so for door, we don't need to download it, or unzip it, so i think we want to fall through all branches on some of the datasets

**Review by akshay-krishnan — APPROVED** [review]

**Comment by ayushbaid** [comment]
I do like the fact that we can run all the benchmarks in parallel but I think we might have overdone it. 

We could have just had 6/7 different bash scripts, one for each benchmark. And from the yml file call the bash script. The bash script downloads the dataset, unzips it, and runs gtsfm for it. The name of the bash script can go in the config.

I feel we have way too many variables which we will change ocassionaly, and hance we can hardcode them in the bash script. No need to have all the names in the zip files too.

For example: a config entry can just be 'palace-of-arts': the bash script will be 'palace-of-arts.sh', and the output zip will be 'palace-of-arts.zip'.

**Comment by johnwlambert** [comment]
> I do like the fact that we can run all the benchmarks in parallel but I think we might have overdone it.
> 
> We could have just had 6/7 different bash scripts, one for each benchmark. And from the yml file call the bash script. The bash script downloads the dataset, unzips it, and runs gtsfm for it. The name of the bash script can go in the config.
> 
> I feel we have way too many variables which we will change ocassionaly, and hance we can hardcode them in the bash script. No need to have all the names in the zip files too.
> 
> For example: a config entry can just be 'palace-of-arts-sift': the bash script will be 'palace-of-arts-sift.sh', and the output zip will be 'palace-of-arts-sift.zip'.

Thanks for the review @ayushbaid. I was trying to avoid having a separate "sift" and "deep" bash script though, since that code would all be copy-pasted. Same for copy-pasting the GDrive download code from one to another. I couldn't figure out how to have a double-for loop in the Github Actions matrix (I don't think it's supported).

How about we make one bash script for gdrive, and then separate one for the other olsson loader ones? I'll make a separate PR for that. I guess in the interest of cutting down the CI time from 5 to 3 hours for all our other PRs, I'll merge this now, and submit the changes to in the next PR.

**Comment by ayushbaid** [comment]
> > I do like the fact that we can run all the benchmarks in parallel but I think we might have overdone it.
> > 
> > We could have just had 6/7 different bash scripts, one for each benchmark. And from the yml file call the bash script. The bash script downloads the dataset, unzips it, and runs gtsfm for it. The name of the bash script can go in the config.
> > 
> > I feel we have way too many variables which we will change ocassionaly, and hance we can hardcode them in the bash script. No need to have all the names in the zip files too.
> > 
> > For example: a config entry can just be 'palace-of-arts-sift': the bash script will be 'palace-of-arts-sift.sh', and the output zip will be 'palace-of-arts-sift.zip'.
> 
> Thanks for the review @ayushbaid. I was trying to avoid having a separate "sift" and "deep" bash script though, since that code would all be copy-pasted. Same for copy-pasting the GDrive download code from one to another. I couldn't figure out how to have a double-for loop in the Github Actions matrix (I don't think it's supported).
> 
> How about we make one bash script for gdrive, and then separate one for the other olsson loader ones? I'll make a separate PR for that. I guess in the interest of cutting down the CI time from 5 to 3 hours for all our other PRs, I'll merge this now, and submit the changes to in the next PR.

I am ok with the current change too! Just one suggestion to improve the names of the artifacts. No need to change it to different scripts.

**Pull request merged**